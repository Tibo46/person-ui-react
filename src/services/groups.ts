import * as t from 'io-ts';
import { get } from '../utils/http';

export const Group = t.strict({
  id: t.string,
  name: t.string,
});
export type Group = t.TypeOf<typeof Group>;

export const GroupResponse = t.array(Group);

export type GroupResponse = t.TypeOf<typeof GroupResponse>;
export async function getTypes(): Promise<Group[]> {
  const res = await get(GroupResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/groups`,
  });

  return res;
}
