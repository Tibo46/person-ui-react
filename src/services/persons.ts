import * as t from 'io-ts';
import { get } from '../utils/http';
import { Group } from './groups';

export const Person = t.strict({
  id: t.string,
  name: t.string,
  group: Group,
});
export type Person = t.TypeOf<typeof Person>;

export const SearchResponse = t.array(Person);

export type SearchResponse = t.TypeOf<typeof SearchResponse>;

interface SearchPersonQuery {
  search: string;
}
export async function searchPersons({ query }: { query: SearchPersonQuery }): Promise<Person[]> {
  const searchTerms = query.search ? `?searchTerms=${query.search}` : '';

  const res = await get(SearchResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/search${searchTerms}`,
  });

  return res;
}

interface GetPersonQuery {
  id: string;
}
export async function getPerson({ query }: { query: GetPersonQuery }): Promise<Person> {
  const res = await get(Person, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/${query.id}`,
  });

  return res;
}
