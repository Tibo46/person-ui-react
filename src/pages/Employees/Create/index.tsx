import React, { useState } from 'react';
import { useAsync } from 'react-async';
import { useHistory } from 'react-router';
import { getGroups } from '../../../services/groups';
import { Person, addPerson } from '../../../services/persons';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import MenuItem from '@material-ui/core/MenuItem';
import Button from '@material-ui/core/Button';
import Loader from '../../../components/Loader';

const CreateEmployee = () => {
  const history = useHistory();
  const [employeeGroup, setEmployeeGroup] = useState(1);
  const [employeeName, setEmployeeName] = useState('');

  const groupsResult = useAsync({
    promiseFn: getGroups,
  });

  async function onSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();

    if (event.currentTarget.checkValidity()) {
      const person: Person = {
        name: employeeName,
        groupId: employeeGroup,
      };

      let newPerson: Person;
      try {
        newPerson = await addPerson(person);
        history.push(`/employees/details/${newPerson.id}`);
      } catch (error) {
        console.error(error);
      }
    }
  }

  return (
    <>
      <Typography variant="h1">Add a new employee</Typography>
      {groupsResult.isLoading ? (
        <Loader />
      ) : groupsResult.data && groupsResult.data.length > 0 ? (
        <form autoComplete="off" className={`container`} onSubmit={onSubmit}>
          <Grid container={true} spacing={2} direction="column">
            <Grid container={true} item={true} spacing={2}>
              <Grid item={true} sm={6}>
                <TextField
                  autoComplete="off"
                  fullWidth={true}
                  required
                  variant="outlined"
                  name="employee-name"
                  label="Employee name"
                  value={employeeName}
                  onChange={(e) => {
                    setEmployeeName(e.target.value);
                  }}
                />
              </Grid>

              <Grid item={true} sm={6}>
                <TextField
                  select
                  fullWidth={true}
                  variant="outlined"
                  name="employee-group"
                  label="Employee group"
                  required
                  value={employeeGroup}
                  onChange={(e) => setEmployeeGroup(parseInt(e.target.value))}
                >
                  {groupsResult.data ? (
                    groupsResult.data?.map((group) => (
                      <MenuItem key={group.id} value={group.id}>
                        {group.name}
                      </MenuItem>
                    ))
                  ) : (
                    <MenuItem key={1} value={0}>
                      Loading...
                    </MenuItem>
                  )}
                </TextField>
              </Grid>
            </Grid>
            <Grid item={true}>
              <Button
                disableRipple={true}
                variant="contained"
                color="primary"
                type="submit"
                size="large"
              >
                Create
              </Button>
            </Grid>
          </Grid>
        </form>
      ) : (
        <Typography className={`no-data`}>
          An error occured and we could not get the Groups. Please try again by refreshing the page.
        </Typography>
      )}
    </>
  );
};

export default CreateEmployee;
