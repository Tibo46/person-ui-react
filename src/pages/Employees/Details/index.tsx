import React from 'react';
import { getPerson } from '../../../services/persons';
import Typography from '@material-ui/core/Typography';
import Loader from '../../../components/Loader';
import { useAsync } from 'react-async';

const DetailsEmployee: React.FC<{ id: string }> = ({ id }) => {
  // @ts-ignore
  const personResult = useAsync({
    promiseFn: getPerson,
    query: {
      id,
    },
    watch: id,
  });

  return (
    <>
      {personResult.isLoading ? (
        <Loader />
      ) : personResult.data ? (
        <>
          <Typography variant="h1">{personResult.data.name}</Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {personResult.data.group?.name}
          </Typography>
        </>
      ) : (
        <Typography className={`no-data`}>This employee does not exists</Typography>
      )}
    </>
  );
};

export default DetailsEmployee;
