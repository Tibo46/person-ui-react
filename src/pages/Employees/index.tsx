import React from 'react';
import { useAsync } from 'react-async';
import Typography from '@material-ui/core/Typography';
import { searchPersons, Person } from '../../services/persons';
import Loader from '../../components/Loader';
import Grid from '@material-ui/core/Grid';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import { TextField, CardMedia, makeStyles, Theme, createStyles } from '@material-ui/core';
import Link from '../../components/Link';
import NoPhoto from '../../images/no-photo.jpg';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    search: {
      marginBottom: '2rem',
    },
    photo: {
      minHeight: '300px',
      objectPosition: 'top',
    },
  })
);

const Employees = () => {
  const classes = useStyles();
  const [filter, setFilter] = React.useState('');

  // @ts-ignore
  const personsResult = useAsync({
    promiseFn: searchPersons,
    query: {
      search: filter,
    },
    watch: filter,
  });

  return (
    <>
      <Typography variant="h1">All your employees</Typography>
      <TextField
        className={classes.search}
        autoComplete="off"
        fullWidth={true}
        variant="outlined"
        label="Start typing to filter employees"
        value={filter}
        name="filter"
        onChange={(e) => {
          setFilter(e.target.value);
        }}
      />

      <div className={`container`}>
        <Grid container={true} spacing={1}>
          {personsResult.isLoading ? (
            <Loader />
          ) : personsResult.data && personsResult.data.length > 0 ? (
            personsResult.data.map((person: Person, index: number) => (
              <Grid item={true} key={`party-${index}`} xl={3} md={4} sm={6} xs={12}>
                <Card>
                  <CardActionArea>
                    <CardMedia
                      className={classes.photo}
                      component="img"
                      alt="Contemplative Reptile"
                      height="140"
                      image={person.photo ? person.photo : NoPhoto}
                      title="Contemplative Reptile"
                    />
                    <Link href={`employees/details/${person.id}`}>
                      <CardContent>
                        <Typography gutterBottom variant="h2">
                          {person.name}
                        </Typography>
                        <Typography variant="body2" color="textSecondary" component="p">
                          {person.group?.name}
                        </Typography>
                      </CardContent>
                    </Link>
                  </CardActionArea>
                </Card>
              </Grid>
            ))
          ) : (
            <Typography className={`no-data`}>
              There is no employee yet. Please add a new one.
            </Typography>
          )}
        </Grid>
      </div>
    </>
  );
};

export default Employees;
