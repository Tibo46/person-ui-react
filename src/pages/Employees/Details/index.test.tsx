import React from 'react';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';

import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import Employees from '.';
import { Person } from '../../../services/persons';
import DetailsEmployee from '.';

const delay = (time = 0) => new Promise((res) => setTimeout(res, time));

window.ENV_DATA = {
  personApiOrigin: 'https://mock-api.com',
} as any;

describe('Employees Details', () => {
  let personResult: Person;

  beforeEach(() => {
    personResult = {
      id: 'employee-1',
      name: 'Walter White',
      groupId: 1,
      group: {
        id: 1,
        name: 'Chemist',
      },
    };
    fetchMock.get(`${window.ENV_DATA.personApiOrigin}/persons/employee-1`, personResult);
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('renders the page', async () => {
    const component = renderer.create(<DetailsEmployee id="employee-1" />);

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should render a loader while fetching data and then display the data', async () => {
    let component: renderer.ReactTestRenderer;
    renderer.act(() => {
      component = renderer.create(<DetailsEmployee id="employee-1" />);
    });

    expect(component!.toJSON()).toMatchSnapshot();

    await renderer.act(async () => {
      await delay();
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should display a loader and then a message saying there is no data, if no data found', async () => {
    fetchMock.get(`${window.ENV_DATA.personApiOrigin}/persons/search`, [], {
      overwriteRoutes: true,
    });

    let component: renderer.ReactTestRenderer;
    renderer.act(() => {
      component = renderer.create(<DetailsEmployee id="employee-not-existing" />);
    });

    expect(component!.toJSON()).toMatchSnapshot();

    await renderer.act(async () => {
      await delay();
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should get the employee data by employee id', async () => {
    await act(async () => {
      const component = mount(<DetailsEmployee id="employee-1" />);
    });

    expect(fetchMock.calls()).toHaveLength(1);
    expect(fetchMock.calls()[0]['identifier']).toBe('https://mock-api.com/persons/employee-1');
  });
});
