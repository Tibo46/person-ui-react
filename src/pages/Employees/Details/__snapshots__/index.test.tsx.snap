// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employees Details renders the page 1`] = `
<div
  className="makeStyles-loader"
/>
`;

exports[`Employees Details should display a loader and then a message saying there is no data, if no data found 1`] = `
<div
  className="makeStyles-loader"
/>
`;

exports[`Employees Details should display a loader and then a message saying there is no data, if no data found 2`] = `
<p
  className="MuiTypography-root no-data MuiTypography-body1"
>
  This employee does not exists
</p>
`;

exports[`Employees Details should render a loader while fetching data and then display the data 1`] = `
<div
  className="makeStyles-loader"
/>
`;

exports[`Employees Details should render a loader while fetching data and then display the data 2`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Walter White
  </h1>,
  <p
    className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
  >
    Chemist
  </p>,
]
`;
